torch_xla/csrc/helpers.cpp

```diff
@@ -525,13 +524,102 @@ std::pair<xla::XlaOp, xla::XlaOp> XlaHelpers::PromoteSecondValue(
-          runtime::util::ToVector<int64_t>(shape2.dimensions())));
+  std::vector<int64_t> dimensions;
+  std::vector<bool> dynamic_dimensions;
+
+  // If the rank of a shape is bigger than then other, fill up the first
+  // dimensions with the ones of the bigger.
+  // Example:
+  //   shape1 = [9, ?, 6, ?, ?]
+  //   shape2 =       [6, 1, 2]
+  // Insert [9, ?] into the dimensions vector.
+  if (shape1.dimensions().size() > shape2.dimensions().size())
+    ExtractDimensionSizesAndDynamicDimensionsFromShape(
+        shape1, shape1.dimensions().size() - shape2.dimensions().size(),
+        dimensions, dynamic_dimensions);
+  else
+    ExtractDimensionSizesAndDynamicDimensionsFromShape(
+        shape2, shape2.dimensions().size() - shape1.dimensions().size(),
+        dimensions, dynamic_dimensions);
+
+  // For the common dimensions, they must match, or one of them be 1.
+  size_t min_size =
+      std::min(shape1.dimensions().size(), shape2.dimensions().size());
+  for (size_t i = 0; i < min_size; i++) {
+    int64_t dim1 =
+        shape1.dimensions()[shape1.dimensions().size() - min_size + i];
+    int64_t dynamic_dim1 =
+        shape1.dynamic_dimensions()[shape1.dynamic_dimensions().size() -
+                                    min_size + i];
+    int64_t dim2 =
+        shape2.dimensions()[shape2.dimensions().size() - min_size + i];
+    int64_t dynamic_dim2 =
+        shape2.dynamic_dimensions()[shape2.dynamic_dimensions().size() -
+                                    min_size + i];
+
+    XLA_CHECK(dim1 == dim2 || dim1 == 1 || dim2 == 1 ||
+              dim1 == xla::Shape::kUnboundedSize ||
+              dim2 == xla::Shape::kUnboundedSize);
+
+    // TODO: Consider replacing the broadcasting logic below with
+    // 'xla::ShapeInference::InferDegenerateDimensionBroadcastShape' resuing the
+    // existing xla shape inference machinery. The function is better suited for
+    // inferring shape when the participating
+    // ranks are the same. However, the function has private visisbilty and
+    // needs some refactoring before use.
+    if (dim1 == 1 || dim2 == 1) {
+      // dim1 | dim2 | result
+      // 1   | X   | X
+      // 1   | <=X | <=X
+      // 1   | ?   | ?
+      // X   | 1   | X
+      // <=X | 1   | <=X
+      // ?   | 1   | ?
+      dimensions.push_back(dim1 == 1 ? dim2 : dim1);
+      dynamic_dimensions.push_back(dim1 == 1 ? dynamic_dim2 : dynamic_dim1);
+    } else if (dim1 == dim2) {
+      // dim1 | dim2 | result
+      // X   | X   | X
+      // X   | <=X | <=X
+      // <=X | X   | <=X
+      // <=X | <=X | <=X
+      // ?   | ?   | ?
+      dimensions.push_back(dim1);
+      dynamic_dimensions.push_back(dynamic_dim1 || dynamic_dim2);
+    } else {
+      // dim1 == xla::Shape::kUnboundedSize || dim2 ==
+      // xla::Shape::kUnboundedSize
+      //
+      // dim1 | dim2 | result
+      // X   | ?   | X
+      // ?   | X   | X
+      // <=X | ?   | ?
```

Comment:
Consider whether `<=X (X=1)` and `? (? = 5)` should infer `<=X` or `?`. The code infers `<=X` and the comment infers `?`. I think it's fine safe to infer `?` that can take any value, but depending on context in PyTorch (and the necessary assertions), the former (current code) may be favorable.

Reply:
You are right! As we discussed offline, the plan is to have a common logic shared between the current logic and https://github.com/openxla/xla/blob/979f5dd0fc0b76bb167c6128a4a6de2ba4c937b9/xla/service/shape_inference.cc#L988. If the expectation from `lhs shape: <=X and rhs_shape: ?` is that the  both should have the same runtime value, then we do not have to consider the special case of X being 1 and ? being 5 at runtime (in comment and in code). I will get back on the that. 
As we are planning to reuse the `ShapeInference::InferDegenerateDimensionBroadcastShape` (after some refactorings), we are planning to  make the fix in the HLO side first.